postgis/lwgeom_in_marc21.c

```diff
@@ -0,0 +1,527 @@
+ *
+ * Copyright (C) 2021 University of Münster (WWU), Germany
+ * Written by Jim Jones <jim.jones@uni-muenster.de>
+ *
+ **********************************************************************/
+
+PG_FUNCTION_INFO_V1(ST_GeomFromMARC21);
+Datum ST_GeomFromMARC21(PG_FUNCTION_ARGS) {
+	GSERIALIZED *geom;
+	LWGEOM *lwgeom;
+	xmlDocPtr xmldoc;
+	text *xml_input;
+	int xml_size;
+	char *xml;
+	xmlNodePtr xmlroot = NULL;
+
+	if (PG_ARGISNULL(0)) PG_RETURN_NULL();
+
+	xml_input = PG_GETARG_TEXT_P(0);
+	xml = text_to_cstring(xml_input);
+	xml_size = VARSIZE_ANY_EXHDR(xml_input);
+
+	xmlInitParser();
+	xmldoc = xmlReadMemory(xml, xml_size, NULL, NULL, XML_PARSE_SAX1);
+
+	if (!xmldoc || (xmlroot = xmlDocGetRootElement(xmldoc)) == NULL) {
+		xmlFreeDoc(xmldoc);
+		xmlCleanupParser();
+		lwpgerror("invalid MARC21/XML document.");
+	}
+
+	lwgeom = parse_marc21(xmlroot);
+
+	xmlFreeDoc(xmldoc);
+	xmlCleanupParser();
+
+	if (lwgeom == NULL) {
+
+		lwgeom_free(lwgeom);
+		PG_RETURN_NULL();
+
+	}
+
+	geom = geometry_serialize(lwgeom);
+
+	lwgeom_free(lwgeom);
+
+	PG_RETURN_POINTER(geom);
+}
+
+static int is_literal_valid(char *literal) {
+
+	int j;
+	int num_dec_sep;
+	int coord_start;
+	int literal_length = strlen(literal);
+
+	POSTGIS_DEBUGF(2, "is_literal_valid called (%s)", literal);
+
+	if (literal_length < 3)	return LW_FALSE;
+
+	coord_start = 0;
+	num_dec_sep = 0;
+
+	/**
+	 * Shortest allowed literal:
+	 *   containing cardinal direction >  cardinal direction + degree with 3 digits
+	 *   no cardinal direction >  degree with 3 digits
+	 *
+	 * The variable "coord_start" stores the position where the coordinates start (0 or 1),
+	 * so that the script can identify where each literal element begins to validate its content.
+	 */
+
+	if (literal[0] == 'N' || literal[0] == 'E' || literal[0] == 'S'	|| literal[0] == 'W' || literal[0] == '+' || literal[0] == '-') {
+
+		if (literal_length < 4) {
+			POSTGIS_DEBUGF(3, "  invalid literal length (%d): \"%s\"", literal_length, literal);
+			return LW_FALSE;
+		}
+
+		coord_start = 1;
+	}
+
+	for (j = coord_start; j < literal_length; j++) {
+
+		if (!isdigit(literal[j])) {
+
+
+			if (j < 3) {
+
+				/**
+				 * The first three characters represent the degrees and they must be numeric.
+				 */
+
+				POSTGIS_DEBUGF(3,"  invalid character '%c' at the degrees section: \"%s\"",	literal[j], literal);
+				return LW_FALSE;
+
+			}
+
+			/**
+			 * Decimal degrees are encoded with after the 3rd character with either a dot or a comma.
+			 * Only one decimal separator is allowed.
+			 */
+			if (literal[j] == '.' || literal[j] == ',') {
+
+				num_dec_sep++;
+
+				if (num_dec_sep > 1) return LW_FALSE;
+
+			} else {
+				POSTGIS_DEBUGF(3, "  invalid character '%c' in %d: \"%s\"",	literal[j], j, literal);
+				return LW_FALSE;
+
+			}
+
+		}
+
+	}
+
+	POSTGIS_DEBUGF(2, "=> is_literal_valid returns LW_TRUE for \"%s\"",	literal);
+	return LW_TRUE;
+
+}
+
+static double parse_geo_literal(char *literal) {
+
+	/**
```

Comment:
There is a *lot* of fiddly parsing logic here, and memcpy calls against math I can't eyeball (which could leave to buffer overwrites, etc, etc) and I wonder: why isn't this just a regex? I mean, learning C regex is a little bit of a trip but it moves a lot of complexity of parsing and string handling into a place where it's not your responsibility anymore. We already have some regex in the code base, and there's an implementation internal to pgsql too which could also be used.

Reply:
Initially I used regex to validate the input strings, but it turned out that it was costing more than 1/3 of the total execution time. Perhaps it was just a bad regex. I'll take another look at regex to try simplify this function.

Reply:
Are you working in such volume that performance implications are worth a complexity trade-off?

Reply:
Well, I've never worked with a data set larger than ~30 million documents, so I guess waiting a bit longer wouldn't hurt that much. I'll try to do some refactoring in this function to make it more palatable (without `memcpy` calls ;) ) before going regex.